File: src/game/state_options.cpp
#include "game/state_options.hpp"

//external includes
#include <CEGUIWindow.h>
#include <CEGUIWindowManager.h>
#include <elements/CEGUIPushButton.h>
#include <elements/CEGUIRadioButton.h>
//custom includes
#include "engine/state_manager.hpp"
#include "engine/settings.hpp"
#include "graphics/context.hpp"

OptionsState::OptionsState()
{
  guiRoot = CEGUI::WindowManager::getSingleton().loadWindowLayout("options.layout");

  ((CEGUI::PushButton*)CEGUI::WindowManager::getSingleton().getWindow("Options/Close"))->subscribeEvent(CEGUI::PushButton::EventClicked, CEGUI::Event::Subscriber(&OptionsState::Close_onClick, this));

  ((CEGUI::RadioButton*)CEGUI::WindowManager::getSingleton().getWindow("Options/GraphicsLow"))->subscribeEvent(CEGUI::RadioButton::EventSelectStateChanged, CEGUI::Event::Subscriber(&OptionsState::Graphics_onStateChange, this));

  ((CEGUI::RadioButton*)CEGUI::WindowManager::getSingleton().getWindow("Options/GraphicsMedium"))->subscribeEvent(CEGUI::RadioButton::EventSelectStateChanged, CEGUI::Event::Subscriber(&OptionsState::Graphics_onStateChange, this));
  
  ((CEGUI::RadioButton*)CEGUI::WindowManager::getSingleton().getWindow("Options/GraphicsHigh"))->subscribeEvent(CEGUI::RadioButton::EventSelectStateChanged, CEGUI::Event::Subscriber(&OptionsState::Graphics_onStateChange, this));

  ((CEGUI::RadioButton*)CEGUI::WindowManager::getSingleton().getWindow("Options/GraphicsExtreme"))->subscribeEvent(CEGUI::RadioButton::EventSelectStateChanged, CEGUI::Event::Subscriber(&OptionsState::Graphics_onStateChange, this));
  
  if(engine::userSettings["graphicsMode"] == "low")
    {
      ((CEGUI::RadioButton*)(CEGUI::WindowManager::getSingleton().getWindow("Options/GraphicsLow")))->setSelected(true);
    }
  else if(engine::userSettings["graphicsMode"] == "medium")
    {
      ((CEGUI::RadioButton*)(CEGUI::WindowManager::getSingleton().getWindow("Options/GraphicsMedium")))->setSelected(true);
    }
  else if(engine::userSettings["graphicsMode"] == "high")
    {
      ((CEGUI::RadioButton*)(CEGUI::WindowManager::getSingleton().getWindow("Options/GraphicsHigh")))->setSelected(true);
    }
  else if(engine::userSettings["graphicsMode"] == "extreme")
    {
      ((CEGUI::RadioButton*)(CEGUI::WindowManager::getSingleton().getWindow("Options/GraphicsExtreme")))->setSelected(true);
    }
}

OptionsState::~OptionsState()
{
  delete guiRoot;
}

void OptionsState::enter()
{
  CEGUI::System::getSingleton().setGUISheet(guiRoot);
  graphics::OgreEngine::getSingleton().setupGUIMgr();
}

void OptionsState::exit()
{
}

void OptionsState::pause()
{
}

void OptionsState::resume()
{
  enter();
}

bool OptionsState::Close_onClick(const CEGUI::EventArgs& args)
{
  engine::StateManager::getSingleton().popState();
  return true;
}

bool OptionsState::Graphics_onStateChange(const CEGUI::EventArgs& args)
{
  unsigned int id = ((CEGUI::RadioButton*)(((CEGUI::WindowEventArgs&)args).window))->getSelectedButtonInGroup()->getID();
  switch(id)
    {
    case 0:
      engine::userSettings["graphicsMode"] = "low";
      break;
    case 1:
      engine::userSettings["graphicsMode"] = "medium";
      break;
    case 2:
      engine::userSettings["graphicsMode"] = "high";
      break;
    case 3:
      engine::userSettings["graphicsMode"] = "extreme";
      break;
    default:
      std::cerr << "Invalid button id: " << id << std::endl;
    }
  engine::saveSettings();
  
  engine::StateManager::getSingleton().onGraphicsSettings();
  return true;
}

File: res/scripts/options/state_options.lua
local CEGUISystem = CEGUI.System:getSingleton()
local CEGUIWindowManager = CEGUI.WindowManager:getSingleton()

class 'OptionsState' (engine.GameState)

function Close_onClick(args)
   StateManager:popState()
   return true
end

function Graphics_onStateChange(args)
   local id = CEGUI.toRadioButton(CEGUI.toWindowEventArgs(args).window):getSelectedButtonInGroup():getID()
   if id==0 then
      engine.setUserSetting("graphicsMode","low")
   end
   if id==1 then
      engine.setUserSetting("graphicsMode","medium")
   end
   if id==2 then
      engine.setUserSetting("graphicsMode","high")
   end
   if id==3 then
      engine.setUserSetting("graphicsMode","extreme")
   end

   StateManager:onGraphicsSettings()
   engine.saveSettings()

   return true
end
   
function OptionsState:getName()
   return "options"
end

local function setRadioButton(name)
   CEGUI.toRadioButton(CEGUIWindowManager:getWindow(name)):setSelected(true)
end   

function OptionsState:__init()
   engine.GameState.__init(self)
   self.guiRoot = CEGUIWindowManager:loadWindowLayout("options.layout")

   CEGUIWindowManager:getWindow("Options/Close"):subscribeEvent("Clicked", "Close_onClick")
   CEGUIWindowManager:getWindow("Options/GraphicsLow"):subscribeEvent("SelectStateChanged", "Graphics_onStateChange")
   CEGUIWindowManager:getWindow("Options/GraphicsMedium"):subscribeEvent("SelectStateChanged", "Graphics_onStateChange")
   CEGUIWindowManager:getWindow("Options/GraphicsHigh"):subscribeEvent("SelectStateChanged", "Graphics_onStateChange")
   CEGUIWindowManager:getWindow("Options/GraphicsExtreme"):subscribeEvent("SelectStateChanged", "Graphics_onStateChange")

   if engine.getUserSetting("graphicsMode")=="low" then
      setRadioButton("Options/GraphicsLow")
   end
   if engine.getUserSetting("graphicsMode")=="medium" then
      setRadioButton("Options/GraphicsMedium")
   end
   if engine.getUserSetting("graphicsMode")=="high" then
      setRadioButton("Options/GraphicsHigh")
   end
   if engine.getUserSetting("graphicsMode")=="extreme" then
      setRadioButton("Options/GraphicsExtreme")
   end
end

function OptionsState:enter()
   CEGUISystem:setGUISheet(self.guiRoot)
   OgreEngine:setupGUIMgr()
end

function OptionsState:exit()
end

function OptionsState:pause()
   self:exit()
end

function OptionsState:resume()
   self:enter()
end
